src/Raven.Pal/src/posix/linuxonly.c

```diff
@@ -108,5 +114,240 @@ rvn_test_storage_durability(
     return rc;
 }
 
+int _io_ring_supported_internal()
+{
+  
+  if(sizeof(void*) != 8)
+    return 0;
 
-#endif
+   struct utsname buffer; 
+   if (uname(&buffer) != 0)
+        return 0;
+
+  int curr_major = 0, curr_minor = 0, curr_patch = 0;
+  sscanf(buffer.release, "%d.%d.%d", &curr_major, &curr_minor, &curr_patch); 
+  
+  return curr_major > 5 || (curr_major==5 && curr_minor>=1);
+}
+
+static bool _io_ring_supported_flag = false;
+static bool _io_ring_supported_cached = false;
+
+bool _io_ring_supported()
+{
+    if(_io_ring_supported_cached)
+        return _io_ring_supported_flag == 1;
+    _io_ring_supported_flag = _io_ring_supported_internal();
+    _io_ring_supported_cached = true;
+    return _io_ring_supported_flag;
+}
+
+
+int32_t _setup_io_ring(struct handle_global_state *global_state, int32_t *detailed_error_code)
+{
+    int rc = io_uring_queue_init(IO_RING_SIZE, &global_state->ring, 0);
+    if(rc)
+    {
+        *detailed_error_code = -rc;
+        return FAIL_CREATE_IO_RING;
+    }
+    return SUCCESS;
+}
+
+void _close_io_ring(struct handle_global_state *global_state)
+{
+    io_uring_queue_exit(&global_state->ring);
+}
+
+
+int32_t _submit_and_wait(
+    struct io_uring* ring,
+    int32_t count,
+    int32_t* detailed_error_code)
+{
+    int32_t rc = io_uring_submit_and_wait(ring, count);
+    if(rc < 0)
+    {
+        *detailed_error_code = -rc;
+        return FAIL_IO_RING_SUBMIT;
+    }
+
+    struct io_uring_cqe* cqe;
+    for(int i = 0; i < count; i++)
+    {
+        rc = io_uring_wait_cqe(ring, &cqe);
+        if (rc < 0)
+        {
+            *detailed_error_code = -rc;
+            return FAIL_IO_RING_NO_RESULT;
+        }
+        if(cqe->res < 0)
+        {
+            *detailed_error_code = -cqe->res;
+            return FAIL_IO_RING_WRITE_RESULT;
+        }
+        io_uring_cqe_seen(ring, cqe);
+    }
+    return SUCCESS;
+}
+
+int32_t rvn_write_io_ring(
+    void *handle,
+    int32_t count,
+    struct page_to_write *buffers,
+    int32_t *detailed_error_code)
+{
+    int32_t rc = SUCCESS;
+    struct handle *handle_ptr = handle;
+    int32_t submitted = 0;
+
+    pthread_mutex_lock(&handle_ptr->global_state->lock);
+
+    for (size_t i = 0; i < count; i++)
+    {
+        struct io_uring_sqe *sqe = io_uring_get_sqe(&handle_ptr->global_state->ring);
+        if (sqe == NULL)
+        {
+            rc = _submit_and_wait(&handle_ptr->global_state->ring, submitted, detailed_error_code);
+            if (rc != SUCCESS)
+            {
+                return rc;
```

Comment:
What is the expected behavior when we fail? We returned error code and never unlocked the mutex. Is there a path to unlock it later?

Reply:
Failure here is the same a fatal error, and will force a restart of the environment.
But *should* release the lock, yes.